### PR TITLE
Authorizer fields

### PR DESCRIPTION
**Pull request by thomasst**
This allows the authorizer to return fields  (e.g. a list of permissions / capabilities / plan type) that are passed to subsequent subscription callbacks.

Questions:
- What should happen when periodic authorization is enabled and the fields change during an active subscription? Should we 1) do nothing (current behavior), 2) have a new specific callback that gets invoked (if defined), 3) re-invoke the `on_subscribe` callback (possibly with a flag that we're resubscribing, or, 4) force an unsubscribe (so the client can re-subscribe).
- Is it necessary to pass authorizer fields to all subsequent subscription callbacks, or is it enough to just pass them to e.g. `before_subscribe` and `on_subscribe`?

### CODE REVIEW

**Comment by thomasst** [comment]
@mpessas any thoughts on the questions?

**Comment by mpessas** [comment]
I think the simplest option would be to pass the authorizer fields with every message. This way, you do not need a separate protocol.

The client would have to verify the capabilities every time.

**Comment by thomasst** [comment]
But, assuming we go with this option:
- With periodic authorization you won't get an immediate notification once the authorizer fields change, i.e. the backend has to wait for the next message from the client. Let's say the client subscribes to updates of an object (but the service should only send updates if certain authorizer fields are set), then the server would keep sending updates potentially forever even if periodic authorization is enabled.
- There is no single callback you can use to validate authorizer fields, you have to validate them in all the message callbacks.

**Comment by jkemp101** [comment]
Seems like something needs to happen when the fields change.  Force unsubscribing seems too severe but might be the easiest.  What happens after re-calling `on_subscribe`, would the client get a message that something changed?

**Comment by thomasst** [comment]
> hat happens after re-calling on_subscribe, would the client get a message that something changed?

Not necessarily. But currently our dialer backend sends a `call` message to the client when subscribing to a dialer and the user is on a call.

Maybe having a separate callback (`on_change` / `on_authorization_change`) is easiest, and implementations can just point it to the same URL as `on_subscribe` if needed.

We can still keep sending all of the authorizer fields to all callbacks like @mpessas suggested.

**Comment by thomasst** [comment]
> There may be something I am missing from the protocol: given a user authenticates (logs in) to a service, how do they log out (destroy the session)? Is this done explicitly somehow or does the session need to be refreshed? In the latter case, the authorization fields re-verification would be part of refreshing the session.

There is no explicit log out command in SocketShark. You could re-auth as somebody else but even that wouldn't automatically invalidate existing subscriptions. Subscriptions are authorized when you subscribe, or re-authorized periodically (if enabled).

**Comment by jkemp101** [comment]
I'm not sure if I understand the big picture here.  I'm getting the feeling we are trying to have socketshark maintain state for other services.  They would trust socketshark to provide their state info like the permissions for a user instead of socketshark providing something like a service session id (which would never change) and let the service restore their own state based on that.  Am I off track here?

**Comment by mpessas** [comment]
> which would never change

That is the problem: things change.

The discussion is around permissions (authorization) and how you verify that a user still has access to a resource. If the session ID never gets invalidated, how do you know that (i) the session is still valid (e.g. you want to revoke it due to a signout event or due to a malicious action) and (ii) the capabilities that accompany the session (permissions) are still valid.

In other words, socketshark creates a session that can go out of sync with the main service's session. The two have to be kept in sync, even if with some delay. Having socketshark periodically validate the session and its capabilities (refresh operation) and passing that information to the service is one solution to the problem.

**Comment by jkemp101** [comment]
My thoughts:
* On field changes a new callback should be called is specified
* I think all subscription callbacks should get these new fields to not limit their use
